File: app/src/main/java/com/example/clojuredroid/TestActivity.java
package com.example.clojuredroid;

import android.graphics.Typeface;
import android.util.Log;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.goodanser.clj_android.runtime.ClojureActivity;

/**
 * Diagnostic activity that validates the Clojure runtime on Android ART
 * and provides interactive nREPL server controls.
 *
 * Extends ClojureActivity, which auto-requires the Clojure namespace
 * com.example.clojuredroid.test-activity and delegates lifecycle calls.
 * Java tests 1-6 validate Java↔Clojure interop; nREPL diagnostics
 * and controls are handled by the Clojure namespace.
 */
public class TestActivity extends ClojureActivity {
    private static final String TAG = "ClojureDroid";

    // Public fields set by test_activity.clj during on-create
    public LinearLayout logLayout;
    public ScrollView scrollView;
    public TextView nreplStatusView;
    public Button startButton;
    public Button stopButton;
    public EditText portInput;

    /**
     * Runs Java↔Clojure interop diagnostic tests 1-6.
     * Returns true if basic runtime tests passed (RT + API accessible),
     * false if a critical failure prevents nREPL from working.
     */
    public boolean runTests() {
        addLog("=== Clojure-Android Diagnostics ===", 0xFFFFFFFF, true);
        addLog("VM: " + System.getProperty("java.vm.name") + " "
                + System.getProperty("java.vm.version"), 0xFFAAAAAA, false);

        // Test 1: Clojure RT initialization
        try {
            long start = System.currentTimeMillis();
            Class<?> rt = Class.forName("clojure.lang.RT");
            long ms = System.currentTimeMillis() - start;
            addLog("1. RT class loaded (" + ms + "ms)", 0xFF00CC00, false);
            addLog("   loader=" + rt.getClassLoader(), 0xFF888888, false);
        } catch (Throwable t) {
            addLog("1. RT class load FAILED: " + t, 0xFFFF0000, false);
            Log.e(TAG, "RT load failed", t);
            return false;
        }

        // Test 2: Clojure API accessible
        try {
            long start = System.currentTimeMillis();
            Object var = clojure.java.api.Clojure.var("clojure.core", "+");
            long ms = System.currentTimeMillis() - start;
            addLog("2. Clojure API accessible (" + ms + "ms)", 0xFF00CC00, false);
        } catch (Throwable t) {
            addLog("2. Clojure API FAILED: " + t, 0xFFFF0000, false);
            Log.e(TAG, "Clojure API failed", t);
            return false;
        }

        // Test 3: Basic Clojure evaluation
        try {
            long start = System.currentTimeMillis();
            Object plus = clojure.java.api.Clojure.var("clojure.core", "+");
            Object result = ((clojure.lang.IFn) plus).invoke(1L, 2L);
            long ms = System.currentTimeMillis() - start;
            boolean ok = "3".equals(result.toString());
            addLog("3. (+ 1 2) = " + result + " (" + ms + "ms)",
                    ok ? 0xFF00CC00 : 0xFFFF0000, false);
        } catch (Throwable t) {
            addLog("3. Basic eval FAILED: " + t, 0xFFFF0000, false);
            Log.e(TAG, "Basic eval failed", t);
        }

        // Test 4: AOT-compiled namespace loading
        try {
            long start = System.currentTimeMillis();
            Object require = clojure.java.api.Clojure.var("clojure.core", "require");
            ((clojure.lang.IFn) require).invoke(
                    clojure.java.api.Clojure.read("com.example.clojuredroid.hello"));
            long ms = System.currentTimeMillis() - start;
            addLog("4. Namespace loaded (" + ms + "ms)", 0xFF00CC00, false);
        } catch (Throwable t) {
            addLog("4. Namespace load FAILED: " + t, 0xFFFF0000, false);
            Log.e(TAG, "Namespace load failed", t);
        }

        // Test 5: Call function from AOT-compiled namespace
        try {
            Object greeting = clojure.java.api.Clojure.var(
                    "com.example.clojuredroid.hello", "greeting");
            Object result = ((clojure.lang.IFn) greeting).invoke();
            addLog("5. greeting = \"" + result + "\"", 0xFF00CC00, false);
        } catch (Throwable t) {
            addLog("5. Function call FAILED: " + t, 0xFFFF0000, false);
            Log.e(TAG, "Function call failed", t);
        }

        // Test 6: Classloader inspection
        try {
            ClassLoader cl = Thread.currentThread().getContextClassLoader();
            String clType = cl != null ? cl.getClass().getName() : "null";
            addLog("6. Context classloader: " + clType, 0xFF00CC00, false);

            // Probe for DynamicClassLoader
            try {
                Class<?> dcl = Class.forName("clojure.lang.DynamicClassLoader");
                addLog("   DynamicClassLoader: present (super="
                        + dcl.getSuperclass().getName() + ")", 0xFF00CC00, false);
            } catch (ClassNotFoundException e) {
                addLog("   DynamicClassLoader: NOT FOUND", 0xFFFF8800, false);
            }

            // Probe for AndroidDynamicClassLoader
            try {
                Class<?> adcl = Class.forName("clojure.lang.AndroidDynamicClassLoader");
                addLog("   AndroidDynamicClassLoader: present (super="
                        + adcl.getSuperclass().getName() + ")", 0xFF00CC00, false);
            } catch (ClassNotFoundException e) {
                addLog("   AndroidDynamicClassLoader: not present (release build)", 0xFFFF8800, false);
            }
        } catch (Throwable t) {
            addLog("6. Classloader check FAILED: " + t, 0xFFFF0000, false);
        }

        return true;
    }

    // ---- UI helpers (called from test_activity.clj) ----

    public android.view.View makeSeparator() {
        android.view.View sep = new android.view.View(this);
        sep.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, 2));
        sep.setBackgroundColor(0xFF444444);
        return sep;
    }

    public void addLog(String text, int color, boolean bold) {
        TextView tv = new TextView(this);
        tv.setText(text);
        tv.setTextSize(13);
        tv.setTextColor(color);
        if (bold) tv.setTypeface(Typeface.DEFAULT_BOLD);
        tv.setPadding(0, 2, 0, 2);
        logLayout.addView(tv);
        // Auto-scroll to bottom
        scrollView.post(() -> scrollView.fullScroll(ScrollView.FOCUS_DOWN));
        if (color == 0xFFFF0000) {
            Log.e(TAG, text);
        } else {
            Log.i(TAG, text);
        }
    }

    /** Log a step from a background thread to both UI and logcat. */
    public void logStep(String msg) {
        Log.d(TAG, msg);
        runOnUiThread(() -> addLog("  " + msg, 0xFF888888, false));
    }

    public void setNreplStatus(String text, int color) {
        if (nreplStatusView != null) {
            nreplStatusView.setText("nREPL: " + text);
            nreplStatusView.setTextColor(color);
        }
    }
}

File: app/src/main/clojure/com/example/clojuredroid/test_activity.clj
(ns com.example.clojuredroid.test-activity
  "Clojure companion for TestActivity. Handles nREPL diagnostics (test 7)
  and interactive nREPL controls, while Java tests 1-6 remain in
  TestActivity.java for intentional Java↔Clojure interop validation."
  (:require [clj-android.repl.server :as repl-server])
  (:import android.graphics.Typeface
           android.text.InputType
           android.view.Gravity
           android.widget.Button
           android.widget.EditText
           android.widget.LinearLayout
           android.widget.LinearLayout$LayoutParams
           android.widget.ScrollView
           android.widget.TextView
           com.example.clojuredroid.TestActivity))

(declare on-start-nrepl on-stop-nrepl)

(def ^:private default-port 7888)

(defn- make-nrepl-panel
  "Builds the nREPL control panel and sets the activity's public fields."
  [^TestActivity activity]
  (let [panel (LinearLayout. activity)]
    (.setOrientation panel LinearLayout/VERTICAL)
    (.setPadding panel 0 16 0 0)

    ;; Status line
    (let [status (TextView. activity)]
      (.setText status "nREPL: checking...")
      (.setTextSize status 16)
      (.setTypeface status Typeface/DEFAULT_BOLD)
      (.setTextColor status (unchecked-int 0xFFCCCC00))
      (.setPadding status 0 8 0 8)
      (set! (.nreplStatusView activity) status)
      (.addView panel status))

    ;; Port row: label + input
    (let [port-row (LinearLayout. activity)]
      (.setOrientation port-row LinearLayout/HORIZONTAL)
      (.setGravity port-row Gravity/CENTER_VERTICAL)

      (let [label (TextView. activity)]
        (.setText label "Port: ")
        (.setTextSize label 16)
        (.setTextColor label (unchecked-int 0xFFCCCCCC))
        (.addView port-row label))

      (let [input (EditText. activity)]
        (.setText input (str default-port))
        (.setInputType input InputType/TYPE_CLASS_NUMBER)
        (.setTextSize input 16)
        (.setMinimumWidth input 200)
        (.setTextColor input (unchecked-int 0xFFFFFFFF))
        (set! (.portInput activity) input)
        (.addView port-row input))

      (.addView panel port-row))

    ;; Button row
    (let [button-row (LinearLayout. activity)]
      (.setOrientation button-row LinearLayout/HORIZONTAL)
      (.setPadding button-row 0 8 0 0)

      (let [start-btn (Button. activity)]
        (.setText start-btn "Start nREPL")
        (.setEnabled start-btn false)
        (.setOnClickListener start-btn
          (reify android.view.View$OnClickListener
            (onClick [_ _] (on-start-nrepl activity))))
        (set! (.startButton activity) start-btn)
        (.addView button-row start-btn))

      (let [stop-btn (Button. activity)]
        (.setText stop-btn "Stop nREPL")
        (.setEnabled stop-btn false)
        (.setOnClickListener stop-btn
          (reify android.view.View$OnClickListener
            (onClick [_ _] (on-stop-nrepl activity))))
        (set! (.stopButton activity) stop-btn)
        (.addView button-row stop-btn))

      (.addView panel button-row))

    panel))

(defn- run-nrepl-diagnostics
  "Runs nREPL diagnostic test 7: checks availability and server status."
  [^TestActivity activity]
  (.addLog activity "" 0 false) ; spacer
  (.addLog activity "=== nREPL Diagnostics ===" (unchecked-int 0xFFFFFFFF) true)

  (if-not (repl-server/repl-available?)
    (do
      (.addLog activity "7. Release build — nREPL infrastructure not available"
               (unchecked-int 0xFFFF8800) false)
      (.setNreplStatus activity "Release build — nREPL not available"
                        (unchecked-int 0xFFFF8800))
      (.setEnabled (.startButton activity) false)
      (.setEnabled (.stopButton activity) false)
      (.setEnabled (.portInput activity) false))
    ;; Debug build — check nREPL resource availability
    (do
      (doseq [cls ["nrepl/server.clj" "nrepl/core.clj" "clj_android/repl/server.clj"]]
        (let [found (some? (.. (class activity) getClassLoader (getResource cls)))]
          (.addLog activity (str "   Resource " cls ": " (if found "found" "MISSING"))
                   (unchecked-int (if found 0xFF00CC00 0xFFFF0000)) false)))

      (.addLog activity "7. clj-android.repl.server loaded" (unchecked-int 0xFF00CC00) false)

      (if (repl-server/running?)
        (let [p (or (repl-server/port) default-port)]
          (.addLog activity (str "   Server already running on port " p " (auto-started by ClojureApp)")
                   (unchecked-int 0xFF00CC00) false)
          (.setNreplStatus activity (str "Running on port " p " (auto-started)")
                           (unchecked-int 0xFF00CC00))
          (.setEnabled (.startButton activity) false)
          (.setEnabled (.stopButton activity) true))
        (do
          (.addLog activity "   Ready — press Start to launch nREPL"
                   (unchecked-int 0xFF00CC00) false)
          (.setNreplStatus activity "Ready — press Start"
                           (unchecked-int 0xFF00CC00))
          (.setEnabled (.startButton activity) true)
          (.setEnabled (.stopButton activity) false))))))

(defn- parse-port [^EditText et]
  (try
    (let [p (Integer/parseInt (.. et getText toString trim))]
      (when (<= 1 p 65535) p))
    (catch NumberFormatException _ nil)))

(defn- on-start-nrepl [^TestActivity activity]
  (let [port (parse-port (.portInput activity))]
    (if-not port
      (.setNreplStatus activity "Invalid port number" (unchecked-int 0xFFFF0000))
      (do
        (.setEnabled (.startButton activity) false)
        (.setEnabled (.portInput activity) false)
        (.setNreplStatus activity (str "Starting nREPL on port " port "...")
                         (unchecked-int 0xFFCCCC00))
        (.addLog activity "" 0 false)
        (.addLog activity (str "Starting nREPL on port " port "...")
                 (unchecked-int 0xFFCCCC00) false)
        (.start
          (Thread.
            (.getThreadGroup (Thread/currentThread))
            (fn []
              (try
                (.logStep activity "Calling (start)...")
                (let [t0 (System/currentTimeMillis)]
                  (repl-server/start port)
                  (let [ms (- (System/currentTimeMillis) t0)
                        actual-port (or (repl-server/port) port)]
                    (.runOnUiThread activity
                      (fn []
                        (.addLog activity
                                 (str "nREPL started on port " actual-port " (" ms "ms)")
                                 (unchecked-int 0xFF00CC00) false)
                        (.addLog activity
                                 (str "Connect: adb forward tcp:" actual-port " tcp:" actual-port)
                                 (unchecked-int 0xFF00CC00) false)
                        (.setNreplStatus activity (str "Running on port " actual-port)
                                         (unchecked-int 0xFF00CC00))
                        (.setEnabled (.stopButton activity) true)))))
                (catch Throwable t
                  (.runOnUiThread activity
                    (fn []
                      (.addLog activity
                               (str "nREPL start FAILED: " (.getName (class t))
                                    ": " (.getMessage t))
                               (unchecked-int 0xFFFF0000) false)
                      (.setNreplStatus activity "Start failed — see log above"
                                       (unchecked-int 0xFFFF0000))
                      (.setEnabled (.startButton activity) true)
                      (.setEnabled (.portInput activity) true))))))
            "nREPL-start"
            1048576))))))

(defn- on-stop-nrepl [^TestActivity activity]
  (.setEnabled (.stopButton activity) false)
  (.setNreplStatus activity "Stopping..." (unchecked-int 0xFFCCCC00))
  (.addLog activity "Stopping nREPL server..." (unchecked-int 0xFFCCCC00) false)
  (.start
    (Thread.
      (fn []
        (try
          (repl-server/stop)
          (.runOnUiThread activity
            (fn []
              (.addLog activity "nREPL server stopped" (unchecked-int 0xFF00CC00) false)
              (.setNreplStatus activity "Stopped — press Start to restart"
                               (unchecked-int 0xFFAAAAAA))
              (.setEnabled (.startButton activity) true)
              (.setEnabled (.portInput activity) true)))
          (catch Throwable t
            (.runOnUiThread activity
              (fn []
                (.addLog activity (str "nREPL stop FAILED: " (.getMessage t))
                         (unchecked-int 0xFFFF0000) false)
                (.setNreplStatus activity "Stop failed" (unchecked-int 0xFFFF0000))
                (.setEnabled (.stopButton activity) true))))))
      "nREPL-stop")))

(defn on-create
  "Called by ClojureActivity when TestActivity is created.
  Builds the layout, runs Java diagnostics 1-6, then handles
  nREPL diagnostics and controls in Clojure."
  [^TestActivity activity _saved-instance-state]
  (let [root (LinearLayout. activity)]
    (.setOrientation root LinearLayout/VERTICAL)
    (.setPadding root 32 32 32 32)

    ;; Scrollable log area
    (let [sv (ScrollView. activity)]
      (.setLayoutParams sv (LinearLayout$LayoutParams.
                             LinearLayout$LayoutParams/MATCH_PARENT
                             0 1.0))
      (let [ll (LinearLayout. activity)]
        (.setOrientation ll LinearLayout/VERTICAL)
        (set! (.logLayout activity) ll)
        (.addView sv ll))
      (set! (.scrollView activity) sv)
      (.addView root sv))

    ;; Separator
    (.addView root (.makeSeparator activity))

    ;; nREPL control panel
    (.addView root (make-nrepl-panel activity))

    (.setContentView activity root)

    ;; Run Java diagnostic tests 1-6
    (let [basic-ok (.runTests activity)]
      (when basic-ok
        ;; Run nREPL diagnostics (test 7) in Clojure
        (run-nrepl-diagnostics activity)))))
